### ISSUE TITLE
Missing resource error messaging

### ISSUE THREAD

**Issue by barclayadam** [pulumi/pulumi#5682]
When re-running one of my stacks I get an error referring to a missing resource, but it's not entirely obvious a) what it is and b) what to do about it.

`error: resource complete event returned an error: failed to verify snapshot: resource urn:pulumi:uat-deploy-uk-uks::s365-outlook-add-in::sig365:DataCentre$sig365:aks:DockerDeployment$kubernetes:apps/v1:Deployment::outlook-addin-api-aks-deployment dependency urn:pulumi:uat-build-global::s365-outlook-add-in::pulumi:pulumi:StackReference::symprex/s365-build-infrastructure/uat-deploy-global refers to missing resource`

I'm not 100% sure on how to read that above message. Is the message saying that `urn:pulumi:uat-build-global::s365-outlook-add-in::pulumi:pulumi:StackReference::symprex/s365-build-infrastructure/uat-deploy-global` is missing, or is _that_ missing a dependency? That `StackReference` does exist.

To fix it I can remove the `urn:pulumi:uat-deploy-uk-uks::s365-outlook-add-in::sig365:DataCentre$sig365:aks:DockerDeployment$kubernetes:apps/v1:Deployment::outlook-addin-api-aks-deployment` resource from the stack export. Re-running the stack then works.

I'm not sure how to reproduce this other than with my exact setup.

OS: windows
Pulumi v2.12.1 using Automation API

**Comment by barclayadam** [comment]
It actually seems the error gets output on the first run through, but the run completes (it actually fails as the container enters a crash loop in Kubernetes).

On subsequent run-through the error stops the stack from running at all

**Comment by barclayadam** [comment]
This issue can, I believe, only happen when using the Automation API. I've not got down to a small repro but the issue is storing a `pulumi.Output` value in one stack and (accidentally) using that in a different stack (both using inline programs).

I do not know if the `StackReference` matters, but the first stack uses a value from a reference in `docker.Image`, then the output is shared from that `docker.Image` to another program (that is the reason for the `StackReference` in the above output).

So this bug report I think becomes to improve the error messaging. Tracking where outputs come from in terms of program and ensuring an early / better error message about this scenario

**Comment by EvanBoyle** [comment]
Nesting Automation API closures, or capturing and sharing raw outputs in the outer scope will not work and is not supported. This should be documented. 

The only supported ways to read values out of a stack is via stack outputs (upRes.outputs).

**Comment by barclayadam** [comment]
Yes that was exactly the problem. In a complex setup it can be quite easy to inadvertently do this which is what happend here.

It all "works", but then fails with the above error message.

This is therefore a feature request for better error messaging around the misuse. Documentation is fine, but it's still easy to get in to this state and the error message does not hint at the problem at all

**Comment by EvanBoyle** [comment]
Yes, the Go SDK checks for nested execution: https://github.com/pulumi/pulumi/blob/master/sdk/go/x/auto/stack.go#L775-L791

Looks like this was missed in the nodejs sdk. Detecting outputs captured in the outer scope probably isn't possible and might require a linter or some external code analysis tool.

**Comment by barclayadam** [comment]
The URN indicates this came from a different project, it is what clued me on to the problem.

Is that not enough to be able to identify this output came from a different project? 

At a minimum perhaps the error message could contain some information about possibly why the problem is happening instead of just being _what_ the problem is
